tracing-appender/src/writer.rs

```diff
@@ -0,0 +1,115 @@
+use crate::rolling::create_writer_file;
+use std::fs::File;
+use std::io;
+
+#[cfg(feature = "compression_gzip")]
+use {
+    crate::compression::CompressionConfig,
+    flate2::write::GzEncoder,
+    std::io::BufWriter,
+    std::sync::{Arc, RwLock},
+};
+
+#[cfg(feature = "compression_gzip")]
+#[derive(Debug)]
+pub(crate) struct CompressedGzip {
+    compression: CompressionConfig,
+    buffer: Arc<RwLock<BufWriter<GzEncoder<BufWriter<File>>>>>,
```

Comment:
a few questions about this:
 - why does this need to be inside an `Arc`? as far as I can tell, we never clone it, so i don't _think_ it needs to be reference counted?
 - i don't love that the `RwLock` here adds a second lock that has to be acquired in the write path when compression is in use, but i see why it's necessary. we probably can't get rid of that without a fairly big rewrite of how the rest of `RollingFileAppender` works, so this is fine for now :/
 - on that subject, I don't think this needs to be a `RwLock`. this is only ever locked mutably to write to it, and there's no case where multiple threads might concurrently access it immutably, so i think this should just be a `Mutex`, which should be a little more lightweight in cases where shared immutable access isn't needed.
 - what's the motivation for wrapping the `GzEncoder` _and_ the `File` itself in separate `BufWriter`s? why do we need two layers of buffering? if this is important, a comment explaining it would be nice.

Reply:
ad 1. Yeah, I might treat `Arc<RwLock<...>>` as an idiom by now.
ad 2. I didn't see another option. I tried for many hours to have it without any sort of lock but I always stumbled on the same problem that the GzEncoder couldn't be changed only behind a `mut ref`.
ad 3. I will change it to Mutex.
ad 4. I will check this again. You are right that it is confusing a bit.

Reply:
> ad 2. I didn't see another option. I tried for many hours to have it without any sort of lock but I always stumbled on the same problem that the GzEncoder couldn't be changed only behind a `mut ref`.

yeah, i think this is fine, since it would probably take a lot of work to work around that part. don't worry about it.